pkg/cmd/pulumi/templates/templates.go

```diff
@@ -0,0 +1,227 @@
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+// Source provides access to a set of project templates, any set of which may be present on
+// disk.
+//
+// Source is responsible for cleaning up old templates, and should always be [Close]d when
+// created.
+type Source struct {
+	templates    []Template
+	errorOnEmpty []error
+	errors       []error
+	closers      []func() error
+	closed       bool
+
+	// m should be held whenever Source is mutated.
+	m  sync.Mutex
+	wg sync.WaitGroup
+}
+
+// Templates lists the templates available to the [Source].
+func (s *Source) Templates() ([]Template, error) {
+	s.wg.Wait() // Wait to ensure that all templates have been fetched before returning the template list.
+
+	s.lockOpen("read templates")
+	defer s.m.Unlock()
+	if err := errors.Join(s.errors...); err != nil {
+		return nil, err
+	}
+	if len(s.templates) == 0 {
+		return nil, errors.Join(s.errorOnEmpty...)
+	}
+	return s.templates, nil
+}
+
+func (s *Source) addTemplate(t Template) {
+	s.lockOpen("add template")
+	s.templates = append(s.templates, t)
+	s.m.Unlock()
+}
+
+func (s *Source) addCloser(f func() error) {
+	s.lockOpen("add closer")
+	s.closers = append(s.closers, f)
+	s.m.Unlock()
+}
+
+func (s *Source) addError(err error) {
+	s.lockOpen("add error")
+	s.errors = append(s.errors, err)
+	s.m.Unlock()
+}
+
+func (s *Source) addErrorOnEmpty(err error) {
+	s.lockOpen("add error on empty")
+	s.errorOnEmpty = append(s.errorOnEmpty, err)
+	s.m.Unlock()
+}
+
+func (s *Source) lockOpen(action string) {
+	s.m.Lock()
+	if s.closed {
+		panic("Attempted to act on closed source: " + action)
+	}
+}
+
+// Close cleans up the [Source] and any associated templates.
+//
+// Close should always be called when [Source] is dropped.
+func (s *Source) Close() error {
+	s.wg.Wait() // Wait to ensure that all templates have been fetched so all closers are visible.
+
+	s.lockOpen("close")
+	defer s.m.Unlock()
+	s.closed = true
+	errs := make([]error, len(s.closers))
+	for i, f := range s.closers {
+		errs[i] = f()
+	}
+	return errors.Join(errs...)
+}
+
+type Template struct {
+	name               string
+	description        string
+	projectDescription string
+
+	error error
+
+	download func(ctx context.Context) (workspace.Template, error)
+
+	source *Source
+}
+
+func (t Template) Name() string               { return t.name }
+func (t Template) Description() string        { return t.description }
+func (t Template) ProjectDescription() string { return t.projectDescription }
+func (t Template) Error() error               { return t.error }
+
+func (t Template) Download(ctx context.Context) (workspace.Template, error) {
+	if t.source == nil {
+		panic("Cannot download a template without a host")
+	}
+	if t.source.closed {
+		panic("Cannot download a template from an already closed host")
+	}
+
+	return t.download(ctx)
+}
+
+type Scope struct{ kind string }
```

Comment:
This needs docs. It's not clear what a scope is, or what the choices below each really mean.